test/dotnet-watch.Tests/HotReload/ApplyDeltaTests.cs

```diff
@@ -67,6 +67,83 @@ public static void Print()
             await App.AssertOutputLineStartsWith("Changed!");
         }
 
+        [Theory]
+        [CombinatorialData]
+        public async Task AutoRestartOnRudeEdit(bool nonInteractive)
+        {
+            var testAsset = TestAssets.CopyTestAsset("WatchHotReloadApp")
+                .WithSource();
+
+            if (!nonInteractive)
+            {
+                testAsset = testAsset
+                    .WithProjectChanges(project =>
+                    {
+                        project.Root.Descendants()
+                            .First(e => e.Name.LocalName == "PropertyGroup")
+                            .Add(XElement.Parse("""
+                                <HotReloadAutoRestart>true</HotReloadAutoRestart>
+                                """));
+                    });
+            }
+
+            var programPath = Path.Combine(testAsset.Path, "Program.cs");
+
+            App.Start(testAsset, nonInteractive ? ["--non-interactive"] : []);
+
+            await App.AssertWaitingForChanges();
+            App.Process.ClearOutput();
+
+            // rude edit: adding virtual method
+            UpdateSourceFile(programPath, src => src.Replace("/* member placeholder */", "public virtual void F() {}"));
+
+            await App.AssertOutputLineStartsWith(MessageDescriptor.WaitingForChanges, failure: _ => false);
+
+            App.AssertOutputContains("⌚ Restart is needed to apply the changes");
+            App.AssertOutputContains($"⌚ [auto-restart] {programPath}(33,11): error ENC0023: Adding an abstract method or overriding an inherited method requires restarting the application.");
+            App.AssertOutputContains($"[WatchHotReloadApp ({ToolsetInfo.CurrentTargetFramework})] Exited");
+            App.AssertOutputContains($"[WatchHotReloadApp ({ToolsetInfo.CurrentTargetFramework})] Launched");
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public async Task AutoRestartOnNoEffectEdit(bool nonInteractive)
+        {
+            var testAsset = TestAssets.CopyTestAsset("WatchHotReloadApp")
+                .WithSource();
+
+            if (!nonInteractive)
+            {
+                testAsset = testAsset
+                    .WithProjectChanges(project =>
+                    {
+                        project.Root.Descendants()
+                            .First(e => e.Name.LocalName == "PropertyGroup")
+                            .Add(XElement.Parse("""
+                            <HotReloadAutoRestart>true</HotReloadAutoRestart>
+                            """));
```

Comment:
nit: indent (like you did above)

```suggestion
                                <HotReloadAutoRestart>true</HotReloadAutoRestart>
                                """));
```